renormalizer/cv/zerot.py

```diff
@@ -0,0 +1,345 @@
+        nonzeros = np.sum(qnmat == constrain_qn)
+
+
+        if self.method == '1site':
+            guess = self.cv_mps[isite - 1][qnmat == constrain_qn].reshape(nonzeros, 1)
+            path_b = [([0, 1], "ab, acd->bcd"),
+                      ([1, 0], "bcd, de->bce")]
+            vec_b = multi_tensor_contract(
+                path_b, second_L, self.b_oper[isite - 1], second_R
+            )[qnmat == constrain_qn].reshape(nonzeros, 1)
+        else:
+            guess = tensordot(
+                self.cv_mps[isite - 2], self.cv_mps[isite - 1], axes=(-1, 0)
+            )
+            guess = guess[qnmat == constrain_qn].reshape(nonzeros, 1)
+            path_b = [([0, 1], "ab, acd->bcd"),
+                      ([2, 0], "bcd, def->bcef"),
+                      ([1, 0], "bcef, fg->bceg")]
+            vec_b = multi_tensor_contract(
+                path_b, second_L, self.b_oper[isite - 2],
+                self.b_oper[isite - 1], second_R
+            )[qnmat == constrain_qn].reshape(nonzeros, 1)
+
+        count = [0]
+        if self.method == "1site":
+            pre_a_mat = einsum('aba, bccd, ede->ace', first_L, self.a_oper[isite - 1],
+                           first_R)[qnmat == constrain_qn]
+        else:
+            pre_a_mat = einsum(
+                'aba, bccd, deef, gfg->aceg', first_L, self.a_oper[isite - 2],
+                self.a_oper[isite - 1], first_R)[qnmat == constrain_qn]
+
+        pre_a_mat = np.diag(1./pre_a_mat.asnumpy())
+
+        def hop(c):
+            count[0] += 1
+            xstruct = solver.cvec2cmat(xshape, c, qnmat, constrain_qn)
+            if self.method == "1site":
+                path_a = [([0, 1], "abc, ade->bcde"),
+                          ([2, 0], "bcde, bdfg->cefg"),
+                          ([1, 0], "cefg, egh->cfh")]
+                ax = multi_tensor_contract(path_a, first_L, Matrix(xstruct),
+                                           self.a_oper[isite - 1], first_R)
+            else:
+                path_a = [([0, 1], "abc, adef->bcdef"),
+                          ([3, 0], "bcdef, bdgh->cefgh"),
+                          ([2, 0], "cefgh, heij->cfgij"),
+                          ([1, 0], "cfgij, fjk->cgik")]
+                ax = multi_tensor_contract(path_a, first_L, Matrix(xstruct),
+                                           self.a_oper[isite - 2], self.a_oper[isite - 1],
+                                           first_R)
+            cout = ax[qnmat == constrain_qn].reshape(nonzeros, 1).asnumpy()
+            return cout
+
+        mat_a = scipy.sparse.linalg.LinearOperator((nonzeros, nonzeros), matvec=hop)
+        if num in [1, 2]:
+            x, info = scipy.sparse.linalg.cg(mat_a, vec_b.asnumpy(), atol=0)
+        else:
+            x, info = scipy.sparse.linalg.cg(mat_a, vec_b.asnumpy(), tol=1.e-5, x0=guess, M=pre_a_mat, atol=0)
+        try:
+            assert info == 0
+        except:
+            print('not convergd')
+            pass
```

Comment:
Just nagging:
```
if info != 0:
    print("not converged")
```
The current version is, of course, fine. You can leave it here.

Reply:
yes, could be better